flashlight/fl/tensor/backend/jit/JitBackend.cpp

```diff
@@ -0,0 +1,537 @@
+    TensorBackend& wrappedBackend,
+    std::function<Tensor(Node*)> jitTensorCreator)
+    : wrappedBackend_(wrappedBackend), jitTensorCreator_(jitTensorCreator) {}
+
+TensorBackendType JitBackend::backendType() const {
+  // Implementers of a backend should create their own option in the
+  // TensorBackendType enum and return it here.
+  return TensorBackendType::Jit;
+}
+
+/* -------------------------- Compute Functions -------------------------- */
+
+void JitBackend::eval(const Tensor& /* tensor */) {
+  // Launch computation for a given tensor. Can be a noop for non-async
+  // runtimes.
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+bool JitBackend::supportsDataType(const fl::dtype& /* dtype */) const {
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+void JitBackend::getMemMgrInfo(
+    const char* /* msg */,
+    const int /* deviceId */,
+    std::ostream* /* ostream */) {
+  // Can be a noop if no memory manager is implemented.
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+void JitBackend::setMemMgrLogStream(std::ostream* /* stream */) {
+  // Can be a noop if no memory manager is implemented.
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+void JitBackend::setMemMgrLoggingEnabled(const bool /* enabled */) {
+  // Can be a noop if no memory manager is implemented.
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+void JitBackend::setMemMgrFlushInterval(const size_t /* interval */) {
+  // Can be a noop if no memory manager is implemented.
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+/* -------------------------- Rand Functions -------------------------- */
+
+void JitBackend::setSeed(const int /* seed */) {
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+Tensor JitBackend::randn(const Shape& /* shape */, dtype /* type */) {
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+Tensor JitBackend::rand(const Shape& /* shape */, dtype /* type */) {
+  FL_JIT_BACKEND_UNIMPLEMENTED;
+}
+
+/* --------------------------- Tensor Operators --------------------------- */
+
+/******************** Tensor Creation Functions ********************/
+#define FL_JIT_BACKEND_CREATE_FUN_LITERAL_DEF_STUB(TYPE)                      \
+  Tensor JitBackend::fromScalar(TYPE /* value */, const dtype /* type */) {   \
+    throw std::invalid_argument(                                              \
+        "JitBackend::fromScalar - not implemented for type " +                \
+        std::string(#TYPE));                                                  \
+  }                                                                           \
+  Tensor JitBackend::full(const Shape& shape, TYPE value, const dtype type) { \
+    switch (type) {                                                           \
+      case dtype::f16:                                                        \
+        return fullWithType(shape, value, dtype::f32).astype(dtype::f16);     \
+      case dtype::f32:                                                        \
+        return fullWithType(shape, value, type);                              \
```

Comment:
Can this function just be
```cpp
if (type == dtype::f16) {
  return fullWithType(shape, value, type).astype(dtype::f16(;
} else {
  return fullWithType(shape, value, type);
}
```
?

Actually: why is it necessary to have this cast with an fp16 type? Can't we just call `fullWithType` with the correctly-propagated value and let the underlying implementation handle the cast?

Reply:
My head was not right when I wrote this...

Yes and I was copying OneDnnBackend (and we couldn't do `fullWithType<float16>`) -- but JIT shouldn't worry about it.

Simplified.